C-Plus-Plus/graphs/Bidirectional_search_using_BFS_in_undirected_graph.cpp

```diff
@@ -0,0 +1,174 @@
+/*
+Bidirectional search is an algorithm in which we find the Smallest Path from the source to the
+target vertex of a graph.
+In this algorithm, we perform the search operation from two sides, i,e. from the source
+vertex(Forward Search) and the target vertex(Backward Search) simultaneously.
+It divides the graph into two smaller sub-graphs, one starting at the source
+vertex and other at the target vertex. 
+As the searching operation proceeds, the two graphs grow and the search terminates when two graphs
+intersect.
+To perform the searching operation, we implement Breadth-first search on both source vertex and
+target vertex.
+
+Bidirectional search is faster in many cases as compared to normal one-directional search as it
+dramatically reduces the amount of required exploration.
+*/
+
+typedef long long int ll;
+#include <iostream>
+#include <vector>
+#include <queue>
+#include <algorithm>
+using namespace std;
+#define mod 1000000007
+
+void bfs(queue<int>& q, vector<bool>& visited, vector<int>& parent, vector<int> graph[])
+{
+    int current = q.front();
+    q.pop();
+
+    for (auto i : graph[current]) {
+        if (!visited[i]) {
+            parent[i] = current;
+            visited[i] = true;
+            q.push(i);
+        }
+    }
+}
+void printShortestPath(vector<int>& parentF, vector<int>& parentB, int sv, int tv, int inode)
+{
+    vector<int> path;
+    path.push_back(inode);
+    int i = inode;
+    while (i != sv) {
+        path.push_back(parentF[i]);
+        i = parentF[i];
+    }
+    reverse(path.begin(), path.end());
+    i = inode;
+    while (i != tv) {
+        path.push_back(parentB[i]);
+        i = parentB[i];
+    }
+
+    for (auto x : path)
+        cout << x << " ";
+    cout << endl;
+}
+
+int bidirectional_search(int sv, int tv, vector<int> graph[], int n, vector<bool>& visitedF,
+    vector<bool>& visitedB, vector<int>& parentF, vector<int>& parentB)
+{
+    queue<int> qF, qB;
+
+    qF.push(sv);
+    visitedF[sv] = true;
+
+    qB.push(tv);
+    visitedB[tv] = true;
+
+    parentF[sv] = -1;
+    parentB[tv] = -1;
+
+    while (!qF.empty() && !qB.empty()) {
+        // Forward Breadth first search
+        bfs(qF, visitedF, parentF, graph); 
+
+        // Backward Breadth first search
+        bfs(qB, visitedB, parentB, graph); 
+
+        int inode = -1;
+        int i, flag = 0;
+        for (i = 0; i < n; i++) {
+            if (visitedF[i] && visitedB[i]) {
+                flag = 1;
+                break;
+            }
+        }
+        if (flag == 1)
+            inode = i;
+
+        if (inode != -1)
+            return inode;
+    }
+    return -1;
+}
+
+int main()

```

Comment:
The main function should act as the driver function, cut short the length of this function by transposing the redundant to a feature/helper function.